### PR TITLE
add feature wasmtime-jitdump

### PR DESCRIPTION
**Pull request by librelois**
Expose wasmtime optional feature jitdump to be able to profile wasm execution with perf.

### CODE REVIEW

File: client/executor/wasmtime/src/runtime.rs [hunk 1]
```diff
@@ -251,6 +251,10 @@ directory = \"{cache_dir}\"
 fn common_config(semantics: &Semantics) -> std::result::Result<wasmtime::Config, WasmError> {
 	let mut config = wasmtime::Config::new();
 	config.cranelift_opt_level(wasmtime::OptLevel::SpeedAndSize);
+	#[cfg(feature = "jitdump")]
+	if std::env::var("WASMTIME_PROFILING_STRATEGY") == Ok("jitdump".to_owned()) {
```
Comment (crystalin) [inline client/executor/wasmtime/src/runtime.rs:255]: Should this env variable be added to https://github.com/paritytech/substrate/blob/master/utils/wasm-builder/src/lib.rs#L125 ?

File: client/executor/wasmtime/Cargo.toml [hunk 2]
```diff
@@ -33,3 +33,6 @@ wasmtime = { version = "0.27.0", default-features = false, features = [
 sc-runtime-test = { version = "2.0.0", path = "../runtime-test" }
 sp-io = { version = "4.0.0-dev", path = "../../../primitives/io" }
 wat = "1.0"
+
+[features]
+jitdump = ["wasmtime/jitdump"]
```
Comment (bkchr) [inline client/executor/wasmtime/Cargo.toml:38]: Why do we need a feature when a env variable needs to be set to have this enabled?

File: client/executor/wasmtime/src/runtime.rs [hunk 3]
```diff
@@ -251,6 +251,10 @@ directory = \"{cache_dir}\"
...
+		config.profiler(wasmtime::ProfilingStrategy::JitDump).unwrap();
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs:256]: No unwraps!

File: client/executor/wasmtime/src/runtime.rs (hunk 1 above)
Comment (bkchr) [reply client/executor/wasmtime/src/runtime.rs:255]: No. It should be documented as part of the crate docs.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: client/executor/wasmtime/Cargo.toml (hunk 2 above)
Comment (librelois) [reply client/executor/wasmtime/Cargo.toml:38]: The idea is to be able to activate or deactivate the profiler without having to recompile the whole project.

File: client/executor/wasmtime/src/runtime.rs (hunk 3 above)
Comment (librelois) [reply client/executor/wasmtime/src/runtime.rs]: done

File: client/executor/wasmtime/Cargo.toml (hunk 2 above)
Comment (bkchr) [reply client/executor/wasmtime/Cargo.toml:38]: I mean, I don't see any reason for having both. I think that only having the env variable should be enough.

Comment (librelois) [reply client/executor/wasmtime/Cargo.toml:38]: If we have only the environment variable it implies that the jitdump feature is always activated when we activate wasmtime, are you ok with that?

File: client/executor/wasmtime/src/runtime.rs (hunk 1 above)
Comment (crystalin) [reply client/executor/wasmtime/src/runtime.rs:255]: Sounds good, can you @librelois add documentation for the crate please ?

File: client/executor/wasmtime/Cargo.toml (hunk 2 above)
Comment (crystalin) [reply client/executor/wasmtime/Cargo.toml:38]: @bkchr I tend to agree with @librelois, I'm not sure the cost of enabling that feature even without setting the profiler. As it is touching a critical part, I want to make sure we don't include useless cost to it.
Your decision Basti

Comment (bkchr) [reply client/executor/wasmtime/Cargo.toml:38]: I just checked the code. The feature is normally enabled by default in wasmtime. As we disabled default features, it is currently not enabled. However, I would assume that this feature isn't influencing the runtime as long as the config option isn't set. 

TLDR: I would enable the feature by default in wasmtime and would not forward this to everywhere. 

@pepyakin WDYT?

Comment (pepyakin) [reply client/executor/wasmtime/Cargo.toml:38]: @bkchr agreed!

Comment (librelois) [reply client/executor/wasmtime/Cargo.toml]: Done

File: client/executor/wasmtime/src/runtime.rs (hunk 1 above)
Comment (librelois) [reply client/executor/wasmtime/src/runtime.rs]: Done

**Review by pepyakin — COMMENTED** [review]
There are a couple of notes:

1. I am all for for the profiler. The jitdump profiler is not that heavy to my knowledge and should not be a problematic to be built by default.
2. I am however not entirely sure about the way it is activated. The configuration IMO should not take place in ad-hoc places like deep in a crate. It should come from the CLI crate (or whatever place is more appropriate in the setting in question). However, I do not intend to block this PR because of that — I realize it is important for the stuff we are doing and block won't be productive.
3. I am not entirely sure that setting some other value should lead to an error. I would prefer to ignore such a value, ideally printing a single warning message. That is not blocking as well.

**Comment by librelois** [comment]
@pepyakin i'm applied 1. and 3. :)
For 2, it would require a lot more changes and we need jitdump quickly, are you ok that we go through an env var for the moment and that the creation of a cli option is for future PR?

File: client/executor/wasmtime/src/runtime.rs [hunk 4]
```diff
@@ -253,6 +253,18 @@ fn common_config(semantics: &Semantics) -> std::result::Result<wasmtime::Config,
...
+		Some(os_string) if os_string == "jitdump" => wasmtime::ProfilingStrategy::JitDump,
+		Some(_) => {
+			log::warn!("WASMTIME_PROFILING_STRATEGY is set to unknown value, ignored.");
+			wasmtime::ProfilingStrategy::None
+		},
+		None => wasmtime::ProfilingStrategy::None,
+	};
+	config
+		.profiler(profiler)
+		.map_err(|e| WasmError::Instantiation(format!("fail to set jitdump profiler: {}", e)))?;
```
Comment (bkchr) [inline client/executor/wasmtime/src/runtime.rs:266]:
```suggestion
		.map_err(|e| WasmError::Instantiation(format!("fail to set profiler: {}", e)))?;
```

**Review by bkchr — APPROVED** [review]
TY

File: client/executor/wasmtime/src/runtime.rs [hunk 5]
```diff
@@ -253,6 +253,18 @@ fn common_config(semantics: &Semantics) -> std::result::Result<wasmtime::Config,
 	config.cranelift_opt_level(wasmtime::OptLevel::SpeedAndSize);
 	config.cranelift_nan_canonicalization(semantics.canonicalize_nans);
 
+	let profiler = match std::env::var_os("WASMTIME_PROFILING_STRATEGY") {
+		Some(os_string) if os_string == "jitdump" => wasmtime::ProfilingStrategy::JitDump,
+		Some(_) => {
+			log::warn!("WASMTIME_PROFILING_STRATEGY is set to unknown value, ignored.");
```
Comment (pepyakin) [inline client/executor/wasmtime/src/runtime.rs:259]: As I said, this should probably be a single message. The reason I am emphasizing this is should this actually be hit, then the user will be drowned in `warn` messages.

**Review by pepyakin — APPROVED** [review]
Otherwise, LGTM.

File: client/executor/wasmtime/src/runtime.rs (hunk 5 above)
Comment (librelois) [reply client/executor/wasmtime/src/runtime.rs]: Ok, to make sure the warning is logged only once I used an AtomicBool, if you prefer another method, please specify :)

Comment (pepyakin) [reply client/executor/wasmtime/src/runtime.rs]: It is fine, although I would put the static declaration right next to the place where we print

Comment (librelois) [reply client/executor/wasmtime/src/runtime.rs]: > I would put the static declaration right next to the place where we print

This is the case, the static is declared just above the function that uses it.

Comment (pepyakin) [reply client/executor/wasmtime/src/runtime.rs]: No, I mean EVEN closer : ) i.e. next to the `if` in the function body.

Comment (librelois) [reply client/executor/wasmtime/src/runtime.rs]: Thanks for the explanation, it's done.
